### PR TITLE
ENH: Add `itkAnalyticSignalImageFilter` and `itkRegionFromReferenceImageFilter` wrappings

### PR DESCRIPTION
**Pull request by tbirdso**
Exposes `itkAnalyticSignalImageFilter` and `itkRegionFromReferenceImageFilter` for illustration in example notebook.

### CODE REVIEW

**Review by dzenanz — APPROVED** [review]

**Comment by tbirdso** [comment]
Seeing a wrapping issue dealing with the `std::complex` type:
> D:\a\im\include\itkAnalyticSignalImageFilter.hxx(219): error C2678: binary '*': no operator found which takes a left-hand operand of type 'std::complex<float>' (or there is no acceptable conversion)
D:\a\im\include\itkAnalyticSignalImageFilter.hxx(219): note: while trying to match the argument list '(std::complex<float>, double)'

Line 219 is trying to multiple a complex and a scalar, which is defined under `<complex>`:
> outputIt.Set(inputIt.Get() * static_cast<typename TInputImage::PixelType>(2));

This issue does not cause the cxx build to fail, only the Python wrapping generation for `AnalyticSignalImageFilter`. I'll also note that this doesn't seem to fail when building with MSBuild, only with GNinja.

@dzenanz @thewtex Are there any additional steps we need to take for wrapping classes dealing with complex numbers? Headers to include?

**Comment by dzenanz** [comment]
> itkBModeImageFilter: warning(4): ITK type not wrapped, or currently not known: itk::FrequencyDomain1DImageFilter< itk::CurvilinearArraySpecialCoordinatesImage< std::complex< double >, 2 >, itk::CurvilinearArraySpecialCoordinatesImage< std::complex< double >, 2 > >

We should use [newer packages](https://github.com/InsightSoftwareConsortium/ITKModuleTemplate/pull/113) for this.

```text
../../../include/itkVnlForward1DFFTImageFilter.hxx: In instantiation of ‘void itk::VnlForward1DFFTImageFilter<TInputImage, TOutputImage>::GenerateData() [with TInputImage = itk::Image<double, 4>; TOutputImage = itk::Image<std::complex<float>, 4>]’:
../../../include/itkVnlForward1DFFTImageFilter.hxx:38:1:   required from here
../../../include/itkVnlForward1DFFTImageFilter.hxx:100:11: error: cannot convert ‘std::complex<double>’ to ‘const PixelType&’ {aka ‘const std::complex<float>&’}
```
To reproduce the error in C++, instantiate `itk::VnlForward1DFFTImageFilter<itk::Image<double, 4>, itk::Image<std::complex<float>, 4>>`. Python wrapping instantiates this, which causes problems.

**Comment by dzenanz** [comment]
Python packages have additional wrappings enabled, compared to defaults. The most recent commit which has made it into 5.2.0.post1:
https://github.com/InsightSoftwareConsortium/ITKPythonPackage/commit/8c0dd734c061969e0b7642b76decb72a710bd5cb
Having these enabled locally should allow you to reproduce the errors.

**Comment by tbirdso** [comment]
Thanks @dzenanz . This will require rebuilding ITK locally with the correct wrappings, right?

**Comment by dzenanz** [comment]
> This will require rebuilding ITK locally with the correct wrappings

Correct.

**Comment by tbirdso** [comment]
After successfully rebuilding ITK I'm seeing new errors in `itkResampleImageFilter.hxx` when building the ITKUltrasound `master` branch:
> C:\repos\ITK\Modules\Filtering\ImageGrid\include\itkResampleImageFilter.hxx(474,20): error C2672: 'itk::CurvilinearArraySpecialCoordinatesImage<PixelType,3>::TransformPhysicalPointToContinuousIndex': no matching overloaded function found
> C:\repos\ITK\Modules\Filtering\ImageGrid\include\itkResampleImageFilter.hxx(475,1): error C2780: 'bool itk::CurvilinearArraySpecialCoordinatesImage<PixelType,3>::TransformPhysicalPointToContinuousIndex(const itk::Point<TCoordRep,3> &,itk::ContinuousIndex<TIndexRep,3> &) const': expects 2 arguments - 1 provided
12>

The line in question in `itkResampleImageFilter.hxx` is here:
>  const auto transformIndex = [outputPtr, transformPtr, inputPtr](const IndexType & index) {
    return inputPtr->template TransformPhysicalPointToContinuousIndex<TInterpolatorPrecisionType>(
      transformPtr->TransformPoint(outputPtr->template TransformIndexToPhysicalPoint<double>(index)));
  };

Given that ITKUltrasound CI is green and ITK built successfully I'm inclined to believe this is a configuration issue on my side. I have `ITK_WRAP_double` and `ITK_WRAP_complex_double` enabled. Are there other configurations that could cause this issue where the base function overload is not found?

**Comment by tbirdso** [comment]
Rebased on `master`. Build+wrap is successful on Linux, still unclear on why my local Windows build is failing but as CI for `master` is passing I don't believe we should consider it a blocking issue.

**Comment by tbirdso** [comment]
`build-linux-python-packages` failure similar to [HASI Issue 56](https://github.com/KitwareMedical/HASI/issues/56).

>  \+ rm dist/itk_bsplinegradient-0.2.4-cp37-cp37m-linux_x86_64.whl
>\+ compgen -G 'dist/itk*-linux*.whl'
~/work/ITKUltrasound/ITKUltrasound
rm: cannot remove './ITKBSplineGradient/doxygen-1.8.11.linux.bin.tar.gz': No such file or directory
Error: Process completed with exit code 1.

Are we calling `rm` in the right directory?

**Comment by tbirdso** [comment]
Rebased on #144.

**Comment by dzenanz** [comment]
It looks like another conflict-resolving rebase is needed.

File: wrapping/itkAnalyticSignalImageFilter.wrap [hunk 1]
```diff
@@ -0,0 +1,10 @@
+itk_wrap_class("itk::AnalyticSignalImageFilter" POINTER)
+  foreach(d ${ITK_WRAP_IMAGE_DIMS})
+    foreach(rt ${WRAP_ITK_REAL})
+      foreach(ct ${WRAP_ITK_COMPLEX_REAL})
```
Comment (thewtex) [inline wrapping/itkAnalyticSignalImageFilter.wrap:4]: >../../../include/itkVnlForward1DFFTImageFilter.hxx:100:24: error: no viable conversion from 'complex<double>' to 'const complex<float>'

Try wrapping only `float` with `complex<float>` and `double` with `complex<double>`

Comment (tbirdso) [reply wrapping/itkAnalyticSignalImageFilter.wrap]: Ah that makes sense. Looks like it fixed warnings on my end, waiting to see what CI says.

**Review by thewtex — APPROVED** [review]
:green_book:

**Pull request merged**